tx/transaction.go

```diff
@@ -83,6 +84,135 @@ type TxData interface {
+func (t *Transaction) MarshalBinary() ([]byte, error) {
+	if t.Type() == TypeLegacy {
+		return rlp.EncodeToBytes(t.body)
+	}
+	var buf bytes.Buffer
+	err := t.encodeTyped(&buf)
+	return buf.Bytes(), err
+}
+
+// encodeTyped writes the canonical encoding of a typed transaction to w.
+func (t *Transaction) encodeTyped(w *bytes.Buffer) error {
+	w.WriteByte(t.Type())
+	return rlp.Encode(w, t.body)
+}
+
+// EncodeRLP implements rlp.Encoder
+func (t *Transaction) EncodeRLP(w io.Writer) error {
+	if t.Type() == TypeLegacy {
+		return rlp.Encode(w, &t.body)
+	}
+	buf := encodeBufferPool.Get().(*bytes.Buffer)
+	defer encodeBufferPool.Put(buf)
+	buf.Reset()
+
+	if err := t.encodeTyped(buf); err != nil {
+		return err
+	}
+
+	return rlp.Encode(w, buf.Bytes())
+}
+
+// UnmarshalBinary decodes the canonical encoding of transactions.
+// It supports legacy RLP transactions and typed transactions.
+func (t *Transaction) UnmarshalBinary(b []byte) error {
+	if len(b) > 0 && b[0] > 0x7f {
+		// It's a legacy transaction.
+		var data legacyTransaction
+		if err := rlp.DecodeBytes(b, &data); err != nil {
+			return err
+		}
+		t.setDecoded(&data, int64(len(b)))
+		return nil
+	}
+	// It's a typed transaction envelope.
+	body, err := t.decodeTyped(b)
+	if err != nil {
+		return err
+	}
+	t.setDecoded(body, int64(len(b)))
+	return nil
+}
+
+// decodeTyped decodes a typed transaction from the canonical format.
+func (t *Transaction) decodeTyped(b []byte) (txData, error) {
+	if len(b) <= 1 {
+		return nil, errShortTypedTx
+	}
+	switch b[0] {
+	case TypeDynamicFee:
+		var body dynamicFeeTransaction
+		err := rlp.DecodeBytes(b[1:], &body)
+		return &body, err
+	default:
+		return nil, ErrTxTypeNotSupported
+	}
+}
+
+// setDecoded sets the inner transaction body and size after decoding.
+func (t *Transaction) setDecoded(body txData, size int64) {
+	t.body = body
+	if size > 0 {
+		t.cache.size.Store(size)
+	}
+}
+
+// DecodeRLP implements rlp.Decoder
+func (t *Transaction) DecodeRLP(s *rlp.Stream) error {
+	kind, size, err := s.Kind()
+
+	switch {
+	case err != nil:
+		return err
+	case kind == rlp.List:
+		// It's a legacy transaction.
+		var body legacyTransaction
+		err = s.Decode(&body)
+		if err == nil {
+			t.setDecoded(&body, int64(rlp.ListSize(size)))
+		}
+
+		return err
+	case kind == rlp.Byte:
+		return errShortTypedTx
+	default:
+		// It's a TX envelope.
+		// First read the tx payload bytes into a temporary buffer.
+		b, err := s.Bytes()
+		if err != nil {
+			return err
+		}
+		body, err := t.decodeTyped(b)
+		if err == nil {
+			t.setDecoded(body, int64(len(b)))
+		}
+		return err
+	}
+}
+
+// Size returns size in bytes when RLP encoded.
+func (t *Transaction) Size() thor.StorageSize {
+	if cached := t.cache.size.Load(); cached != 0 {
+		return thor.StorageSize(cached)
+	}
+
+	var size thor.StorageSize
+	rlp.Encode(&size, t)
+
+	// For typed transactions, the encoding also includes the leading type byte.
+	if t.body.txType() != TypeLegacy {
```

Comment:
I can see where the byte is written when encoding, so I assume `Size` was not working correctly before this PR, right?

Reply:
Yes, I think so.